### PR TITLE
API changes - separate calls etc. Replaces #276.

### PR DESCRIPTION
**Pull request by levb**
**REPLACES:** https://github.com/mattermost/mattermost-plugin-apps/pull/276

#### Summary
Add separate calls for each function on Apps (submit, form source, form submit and lookup). Also updated the filtering with the new logic.

This PR also piggybacks the renaming of markdown and error to text, and filtering out "white space labels" (to avoid no text labels on post options and channel header).

Notes for testing: This is a breaking change. All Apps will need to be updated to work properly. The test app will be updated to fit the new approach.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-39092
https://mattermost.atlassian.net/browse/MM-39096
https://mattermost.atlassian.net/browse/MM-31121
https://mattermost.atlassian.net/browse/MM-33532

#### Related Pull Requests
Webapp: https://github.com/mattermost/mattermost-webapp/pull/9263
Mobile: https://github.com/mattermost/mattermost-mobile/pull/5877
Test App: https://github.com/mattermost/mattermost-app-test/pull/14
ServiceNow: https://github.com/mattermost/mattermost-app-servicenow/pull/37

### Breaking Changes
Call API, routing

#### TODO
The following apps need to get updated:
- [x] Examples
- [x] Test app
- [x] ServiceNow
- [ ] Zendesk
- [x] gCal

### CODE REVIEW

File: cmd/appsctl/aws.go [hunk 1]
```diff
@@ -235,9 +235,7 @@ var awsTestLambdaCmd = &cobra.Command{
 		}
 
 		creq := apps.CallRequest{
-			Call: apps.Call{
-				Path: "/ping",
-			},
+			Call: *apps.NewCall("/ping"),
```
Comment (hanzei) [inline cmd/appsctl/aws.go:238]: Does a builder pattern as described [here](https://devcharmander.medium.com/design-patterns-in-golang-the-builder-dac468a71194) make more sense for constructing calls? That way we should get rid of the de-referencing.

File: utils/httputils/utils.go [hunk 2]
```diff
@@ -85,35 +85,29 @@ func WriteJSON(w http.ResponseWriter, v interface{}) error {
...
-	return HandleData("application/json", data)
+func HandleStaticJSONData(data []byte) http.HandlerFunc {
+	return HandleStaticData("application/json", data)
 }
 
-// HandleData returns an http.HandleFunc that serves a data chunk with a
+// HandleStaticData returns an http.HandleFunc that serves a data chunk with a
 // specified content-type.
-func HandleData(ct string, data []byte) http.HandlerFunc {
+func HandleStaticData(ct string, data []byte) http.HandlerFunc {
```
Comment (hanzei) [inline utils/httputils/utils.go:96]: The `static` part of the method confuses me. The method handles the data passed to it. `static` let's me think that the method does the name interdependently of it's arguments, which is not the case.

File: server/httpin/restapi/call.go [hunk 3]
```diff
@@ -44,15 +44,6 @@ func (a *restapi) Call(w http.ResponseWriter, req *http.Request, in proxy.Incomi
...
 
-	a.conf.Logger().Debugw(
-		"Received call response",
-		"app_id", creq.Context.AppID,
-		"acting_user_id", in.ActingUserID,
-		"error", res.ErrorText,
-		"type", res.Type,
-		"path", creq.Path,
-	)
-
```
Comment (hanzei) [inline server/httpin/restapi/call.go:55]: Why was this removed?

File: server/proxy/bindings.go [hunk 4]
```diff
@@ -86,122 +91,156 @@ func (p *Proxy) GetAppBindings(in Incoming, cc apps.Context, app apps.App) []app
...
+	conf config.Config,
+) (*apps.Binding, []error) {
+	var problems []error
+
+	// Cleanup Location.
+	if b.Location == "" {
+		b.Location = apps.Location(app.Manifest.AppID)
+	}
+	if trimmed := apps.Location(strings.TrimSpace(string(b.Location))); trimmed != b.Location {
+		problems = append(problems, errors.Errorf("trimmed whitespace from location %s", trimmed))
```
Comment (hanzei) [inline server/proxy/bindings.go:157]: Is that really worth throwing an error for?

File: server/proxy/bindings.go [hunk 5]
```diff
@@ -86,122 +91,156 @@ func (p *Proxy) GetAppBindings(in Incoming, cc apps.Context, app apps.App) []app
...
+		b.Location = trimmed
+	}
 
-			locationsUsed[b.Location] = true
-			labelsUsed[b.Label] = true
-			b.AppID = app.Manifest.AppID
+	fql := locPrefix.Sub(b.Location)
+	allowed := false
+	for _, grantedLoc := range app.GrantedLocations {
+		// TODO Why `grantedLoc.In(fql)`?
```
Comment (hanzei) [inline server/proxy/bindings.go:164]: @larkox :point_up:

File: server/proxy/bindings.go [hunk 6]
```diff
@@ -86,122 +91,156 @@ func (p *Proxy) GetAppBindings(in Incoming, cc apps.Context, app apps.App) []app
...
+	if fql != apps.LocationCommand && fql.In(apps.LocationCommand) {
+		// A command binding must have a valid label. Default to Location if needed.
+		if b.Label == "" {
+			b.Label = string(b.Location)
+		}
+		if trimmed := strings.TrimSpace(b.Label); trimmed != b.Label {
+			problems = append(problems, errors.Errorf("trimmed whitespace from label %s", trimmed))
+			b.Label = trimmed
 		}
+		if strings.ContainsAny(b.Label, " \t") {
```
Comment (hanzei) [inline server/proxy/bindings.go:190]: Does this catch all cases? What about `foo bar`?

File: server/proxy/bindings.go [hunk 7]
```diff
@@ -86,122 +91,156 @@ func (p *Proxy) GetAppBindings(in Incoming, cc apps.Context, app apps.App) []app
...
-			if locationsUsed[b.Location] || labelsUsed[b.Label] {
-				continue
-			}
+func cleanAppBinding(
+	app apps.App,
+	b apps.Binding,
+	locPrefix apps.Location,
+	userAgent string,
+	conf config.Config,
+) (*apps.Binding, []error) {
```
Comment (hanzei) [inline server/proxy/bindings.go:149]: Should we use https://pkg.go.dev/github.com/hashicorp/go-multierror here as it stick with the standard of returning `error` and it's already used in `apps`?

**Review by hanzei — COMMENTED** [review]
Great simplification :+1:

File: apps/call.go [hunk 8]
```diff
@@ -47,163 +28,65 @@ type Call struct {
...
-	}
-}
-
-func NewDataResponse(data interface{}) CallResponse {
-	return CallResponse{
-		Type: CallResponseTypeOK,
-		Data: data,
+func (c *Call) UnmarshalJSON(data []byte) error {
+	stringValue := ""
+	err := json.Unmarshal(data, &stringValue)
```
[the remark this answers is not in the document]
Comment (levb) [reply apps/call.go:33]: 1/5 it's simpler if you write JSON or even TS once the TS classes are updated. For golang, this is of no value.

File: cmd/appsctl/aws.go (hunk 1 above)
Comment (levb) [reply cmd/appsctl/aws.go:238]: Agree, but perhaps a separate (HW?) ticket?

File: server/httpin/restapi/call.go (hunk 3 above)
Comment (levb) [reply server/httpin/restapi/call.go:55]: I found this particular log to be useless, spammy, and verbose on Text responses.

File: server/proxy/bindings.go (hunk 4 above)
Comment (levb) [reply server/proxy/bindings.go:157]: We still include the binding, and yes it is "unclean" if it has whitespace like that.

File: server/proxy/bindings.go (hunk 6 above)
Comment (levb) [reply server/proxy/bindings.go:190]: @hanzei please elaborate? 1/5 We could do regexp `\s`, but seems an overkill, ` ` and `\t` seem good enough?

File: server/builtin/debug_kv_edit_modal.go [hunk 9]
```diff
@@ -13,135 +13,123 @@ import (
...
-		},
+			}))
 
-		submitf: func(creq apps.CallRequest) apps.CallResponse {
-			action := creq.GetValue(fAction, "")
-			newValue := creq.GetValue(fNewValue, "")
-			key, _ := creq.State.(string)
-			loc := a.newLocalizer(creq)
+	default:
+		return apps.NewErrorResponse(errors.New("don't know what to do: %q"))
```
Comment (mickmister) [inline server/builtin/debug_kv_edit_modal.go:58]:
```suggestion
		return apps.NewErrorResponse(errors.Errorf("don't know what to do: %q", action))
```

**Review by mickmister — APPROVED** [review]
LGTM :tada:

Just a few questions

File: cmd/appsctl/aws.go (hunk 1 above)
Comment (hanzei) [reply cmd/appsctl/aws.go:238]: Sound good to me. Would you please file one?

File: server/httpin/restapi/call.go (hunk 3 above)
Comment (hanzei) [reply server/httpin/restapi/call.go:55]: 2/ I kind of liked it as it give a clear notice that and what response the apps plugin got

File: server/proxy/bindings.go (hunk 6 above)
Comment (hanzei) [reply server/proxy/bindings.go:190]: I didn't saw the white space before `\t`. LGTM :+1:

File: server/proxy/bindings.go (hunk 5 above)
Comment (larkox) [reply server/proxy/bindings.go:164]: The main logic is that if the granted location is `/command`, you want to make sure `/command` is part of your location (e.g. `/command/myApp/subscribe`). What I don't remember is why we also use `fql.In(grantedLoc)`. It should be enough with just `grantedLoc.In(fql)`.

**Review by larkox — APPROVED** [review]
Apart of what has been already said, LGTM.

File: cmd/appsctl/aws.go (hunk 1 above)
Comment (levb) [reply cmd/appsctl/aws.go:238]: Filed https://mattermost.atlassian.net/browse/MM-40508

File: apps/call.go [hunk 10]
```diff
@@ -47,163 +28,65 @@ type Call struct {
...
-func CallRequestFromJSONReader(in io.Reader) (*CallRequest, error) {
-	c := CallRequest{}
-	err := json.NewDecoder(in).Decode(&c)
-	if err != nil {
-		return nil, err
-	}
-	return &c, nil
+func (c Call) WithState(state interface{}) *Call {
+	c.State = state
+	return &c
```
[the remark this answers is not in the document]
Comment (levb) [reply apps/call.go:74]: A pointer can always be taken a value of, https://go.dev/play/p/wN_rZMKB5Hp

File: apps/call.go [hunk 11]
```diff
@@ -236,58 +119,12 @@ func (c *Call) PartialCopy() *Call {
...
-	}
-
-	if b, ok := isBool(creq.Values[name]); ok {
-		return b
-	}
-	opt, ok := creq.Values[name].(map[string]interface{})
-	if ok {
-		if v, ok2 := isBool(opt["value"]); ok2 {
-			return v
+	if state, ok := clone.State.(map[string]string); ok {
```
[the remark this answers is not in the document]
Comment (levb) [reply apps/call.go:122]: It doesn't, but PartialCopy only copies these 2 map types

File: server/builtin/field_app_id.go [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+			ID:    "field.appID.description",
+			Other: "Select an App or enter the App ID",
+		}),
+		Label: a.conf.I18N().LocalizeDefaultMessage(loc, &i18n.Message{
+			ID:    "field.appID.label",
+			Other: "app",
+		}),
+		AutocompletePosition: autocompletePos,
+		IsRequired:           isRequired,
+		SelectDynamicLookup:         lookupCall,
```
[the remark this answers is not in the document]
Comment (levb) [reply server/builtin/field_app_id.go:38]: `WARN [runner] The linter 'golint' is deprecated (since v1.41.0) due to: The repository of the linter has been archived by the owner.  Replaced by revive.`

@hanzei you know anything about this?

File: server/builtin/debug_kv_list.go [hunk 13]
```diff
@@ -9,141 +9,92 @@ import (
...
+			},
 		})
 	}
-	return options, nil
+	message += "\n"
+
+	if encode {
+		message += a.conf.I18N().LocalizeDefaultMessage(loc, &i18n.Message{
+			ID:    "command.debug.kv.list.submit.note",
+			Other: "**NOTE**: keys are base64-encoded for pasting into `/apps debug kv edit` command. Use `/apps debug kv list --base64 false` to output raw values.",
```
[the remark this answers is not in the document]
Comment (levb) [reply server/builtin/debug_kv_list.go:87]: apps-only, but we can write a (plugin) apop for debugging KV store. Or... we can get a better KV store.

File: apps/call.go [hunk 14]
```diff
@@ -47,163 +28,65 @@ type Call struct {
...
-	c := CallRequest{}
-	err := json.Unmarshal(data, &c)
-	if err != nil {
-		return nil, err
-	}
-	return &c, nil
+func (c Call) WithExpand(expand Expand) *Call {
+	c.Expand = &expand
+	return &c
 }
```
Comment (hanzei) [inline apps/call.go:70]: Can we simplify this to 
```suggestion
func (c Call) WithExpand(expand Expand) Call {
	c.Expand = &expand
	return c
}
```
?

File: server/builtin/field_app_id.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply server/builtin/field_app_id.go:38]: The go-to version for `golangci-lint` is v1.33.0 right now. That one should complain about the issue above

**Review by hanzei — APPROVED** [review]

**Comment by hanzei** [comment]
@levb CI is complaining about 'server/httpin/restapi/restapitestlib.go:114:41: undefined: httputils.HandleStaticJSON`

File: server/builtin/field_app_id.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (levb) [reply server/builtin/field_app_id.go:38]: I installed 1.33 and it does not appear to be catching `gofmt` issues either

File: apps/call.go (hunk 14 above)
Comment (levb) [reply apps/call.go:70]: returning a *Call was intentional since a lot of the internal functions and structs take a *Call, and you can't do `&NewCall()`

File: utils/httputils/utils.go [hunk 15]
```diff
@@ -85,35 +85,29 @@ func WriteJSON(w http.ResponseWriter, v interface{}) error {
...
-	return HandleData("application/json", data)
+func DoHandleJSONData(data []byte) http.HandlerFunc {
+	return DoHandleData("application/json", data)
 }
 
-// HandleData returns an http.HandleFunc that serves a data chunk with a
+// DoHandleData returns an http.HandleFunc that serves a data chunk with a
 // specified content-type.
-func HandleData(ct string, data []byte) http.HandlerFunc {
+func DoHandleData(ct string, data []byte) http.HandlerFunc {
```
Comment (hanzei) [inline utils/httputils/utils.go:96]: 0/5 Why the `Do` prefix?

Comment (levb) [reply utils/httputils/utils.go:96]: Because `Handle...` is to registyer handlers, by existing convention

File: server/builtin/field_app_id.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (hanzei) [reply server/builtin/field_app_id.go]: Yep, can confirm that. I don't know why this happens. Seems to be a bug in `golangci-lint`.

File: server/proxy/bindings.go (hunk 5 above)
Comment (levb) [reply server/proxy/bindings.go]: See newly added test, should be only `fql.In(grantedLoc)`

**Comment by levb** [comment]
@DHaussermann this should be testable with https://github.com/mattermost/mattermost-app-test/pull/14

File: apps/field.go [hunk 16]
```diff
@@ -76,7 +76,7 @@ type Field struct {
 	AutocompletePosition int `json:"position,omitempty"`
 
 	// TODO: ModalLabel should default to Label, Name
```
Comment (hanzei) [inline apps/field.go:78]: Is this still a TODO?

File: apps/app.go [hunk 17]
```diff
@@ -93,6 +93,7 @@ func DecodeCompatibleApp(data []byte) (app *App, err error) {
 // It is used to describe the OAuth2 connections both to Mattermost, and
 // optionally to a 3rd party remote system.
 type OAuth2App struct {
+	RemoteURL    string `json:"remote_url,omitempty"`
```
Comment (levb) [inline apps/app.go:96]: 0/5 `RemoteRootURL`?

File: apps/call_request.go [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+func CallRequestFromJSONReader(in io.Reader) (*CallRequest, error) {
+	c := CallRequest{}
+	err := json.NewDecoder(in).Decode(&c)
+	if err != nil {
+		return nil, err
+	}
+	return &c, nil
+}
+
+func (creq *CallRequest) GetValue(name, defaultValue string) string {
```
Comment (levb) [inline apps/call_request.go:77]: It appears convenient to have a function that looks up in `Values`->`State`->default.

@hanzei @mickmister Do you think we should have `GetString`/`GetBool[Value]` with the ^^ semantics defined on `CallRequest`?

File: examples/go/hello-serverless/aws/go.mod [hunk 19]
```diff
@@ -5,7 +5,7 @@ go 1.16
 require (
 	github.com/aws/aws-lambda-go v1.19.1
 	github.com/awslabs/aws-lambda-go-api-proxy v0.11.0
-	github.com/mattermost/mattermost-plugin-apps v0.7.1-0.20211011164942-ce8ff2bc1616
+	github.com/mattermost/mattermost-plugin-apps v0.7.1-0.20211207113443-6e904255aa08
```
Comment (levb) [inline examples/go/hello-serverless/aws/go.mod:8]: check and update?

Comment (levb) [reply examples/go/hello-serverless/aws/go.mod:8]: Also, maybe try a single `go.mod` now again, with kubeless gone?

File: examples/go/hello-serverless/function/go.mod [hunk 20]
```diff
@@ -2,4 +2,4 @@ module github.com/mattermost/mattermost-plugin-apps/examples/go/hello-serverless
 
 go 1.16
 
-require github.com/mattermost/mattermost-plugin-apps v0.7.1-0.20211011164942-ce8ff2bc1616
+require github.com/mattermost/mattermost-plugin-apps v0.7.1-0.20211207113443-6e904255aa08
```
Comment (levb) [inline examples/go/hello-serverless/function/go.mod:5]: a single go.mod at the top?

File: apps/call_request.go (hunk 18 above)
Comment (levb) [reply apps/call_request.go:77]: Filed https://mattermost.atlassian.net/browse/MM-42214

File: examples/go/hello-serverless/aws/go.mod (hunk 19 above)
Comment (levb) [reply examples/go/hello-serverless/aws/go.mod:8]: treat as NS, leaving for the doc round of examples cleanup

File: apps/app.go (hunk 17 above)
Comment (levb) [reply apps/app.go]: done, will affect [ServiceNow](https://github.com/mattermost/mattermost-app-servicenow/pull/37)
